sky/spot/core.py

```diff
@@ -0,0 +1,299 @@
+    # The stderr is redirected to stdout
+    returncode, stdout, _ = backend.run_on_head(handle,
+                                                code,
+                                                require_outputs=True,
+                                                stream_logs=False)
+    try:
+        subprocess_utils.handle_returncode(returncode, code,
+                                           'Failed to cancel managed spot job',
+                                           stdout)
+    except exceptions.CommandError as e:
+        raise RuntimeError(e.error_msg) from e
+
+    sky_logging.print(stdout)
+    if 'Multiple jobs found with name' in stdout:
+        with ux_utils.print_exception_no_traceback():
+            raise RuntimeError(
+                'Please specify the job ID instead of the job name.')
+
+
+@usage_lib.entrypoint
+def tail_logs(name: Optional[str], job_id: Optional[int], follow: bool) -> None:
+    # NOTE(dev): Keep the docstring consistent between the Python API and CLI.
+    """Tail logs of managed spot jobs.
+
+    Please refer to the sky.cli.spot_logs for the document.
+
+    Raises:
+        ValueError: invalid arguments.
+        sky.exceptions.ClusterNotUpError: the spot controller is not up.
+    """
+    # TODO(zhwu): Automatically restart the spot controller
+    handle = backend_utils.is_controller_accessible(
+        controller_type=controller_utils.Controllers.SPOT_CONTROLLER,
+        stopped_message=('Please restart the spot controller with '
+                         f'`sky start {spot_utils.SPOT_CONTROLLER_NAME}`.'))
+
+    if name is not None and job_id is not None:
+        raise ValueError('Cannot specify both name and job_id.')
+    backend = backend_utils.get_backend_from_handle(handle)
+    assert isinstance(backend, backends.CloudVmRayBackend), backend
+    # Stream the realtime logs
+    backend.tail_spot_logs(handle, job_id=job_id, job_name=name, follow=follow)
+
+
+@usage_lib.entrypoint
+def launch(
+    task: Union['sky.Task', 'sky.Dag'],
+    name: Optional[str] = None,
+    stream_logs: bool = True,
+    detach_run: bool = False,
+    retry_until_up: bool = False,
+):
+    # NOTE(dev): Keep the docstring consistent between the Python API and CLI.
+    """Launch a managed spot job.
+
+    Please refer to the sky.cli.spot_launch for the document.
+
+    Args:
+        task: sky.Task, or sky.Dag (experimental; 1-task only) to launch as a
+          managed spot job.
+        name: Name of the spot job.
+        detach_run: Whether to detach the run.
+
+    Raises:
+        ValueError: cluster does not exist.
+        sky.exceptions.NotSupportedError: the feature is not supported.
+    """
+    entrypoint = task
+    dag_uuid = str(uuid.uuid4().hex[:4])
+
+    dag = dag_utils.convert_entrypoint_to_dag(entrypoint)
+    assert dag.is_chain(), ('Only single-task or chain DAG is '
+                            'allowed for spot_launch.', dag)
```

Comment:
code gardening: should we raise errors instead of assert to give more information to the user

Reply:
Good point! Updated to raising ValueError instead.